### PR TITLE
Added new intrinsic support.

### PR DESCRIPTION
**Pull request by ghost**
Added support for:
1. Aesni
2. sha256
3. Rdrand
4. rdseed
5. sha1 (Almost complete. I don't believe the Intrinsic structure fully supports llvm.x86.sha1rnds4)

I added several SSE codes. As well as the SSE2 intrinsic for cache flushing. There were a few bugs with the SSE2 memory fenses. Sfense doesn't exist. Mfense was pointing to llvm.x86.sse2.fense which doesn't exist.

### CODE REVIEW

**Comment by bluss** [comment]
Hi! Thanks for the PR!

Unfortunately platform intrinsics are not added this way, the whole file is generated by a python script, [see this comment.](https://github.com/valarauca/rust/blob/master/src/librustc_platform_intrinsics/x86.rs#L11)

**Comment by alexcrichton** [comment]
Ah yes @bluss is right, can you modify the files that these are generated from and then use them to regenerate the file? The `src/etc/platform-intrinsics` directory should have directions.

**Comment by valarauca** [comment]
So whats the process for adding fma support? Would the python script handle adding new json files, or would that require modification of the python script? Wondering for the AES/SHA1/SHA256 extensions.

**Comment by alexcrichton** [comment]
cc @ruud-v-a (you've been working on this as well)

I'm not entirely sure as to the exact syntax itself, but if you modify one of the `x86/*.json` files and then run the script again it should be enough?

**Comment by valarauca** [comment]
I'm not 100% sure. My python is super rusty. But the intrinsic I patched in have a different prefix. Which the file are organized by pre-ix, and judging from the JSON the fields are format qualifiers which may include the llvm prefix that they're grouped with.

To get further into it a decent chunk of intrinsic that were added with SSE2 aren't actually listed under the llvm.x86.sse2.* prefix, but have their own prefix. (This can be repeated for every intrinic set). [1]

Basically I have no clue how to submit this patch via the json file. The info.json is incomplete (I'm assuming its an example) as it lists some data types, but doesn't explain others. [2]

[1] I think this can be avoided with the ! qualifier, but idk. 

[2] Example: https://github.com/rust-lang/rust/blob/c19e7b629b42fc2e153893762397a336423e6ec3/src/etc/platform-intrinsics/x86/avx.json#L100

**Comment by ruud-v-a** [comment]
The generator and syntax in the json are documented [here](https://github.com/rust-lang/rust/blob/master/src/etc/platform-intrinsics/generator.py#L564). To regenerate the rs files I used the following command:

    $ cd src/etc/platform-intrinsics/x86
    $ python2 ../generator.py --format compiler-defs -i info.json   \
      sse.json sse2.json sse3.json ssse3.json sse41.json sse42.json \
      avx.json avx2.json fma.json                                   \
      > ../../../librustc_platform_intrinsics/x86.rs

So the generator does not know which files to include, you have to pass them to the script. (Or I missed something.)

There was no file for FMA before, but I added it and now it is there. (Only the `_ss` and `_sd` intrinsics are still missing, the others are actually part of AVX-512 and KNC according to the Intel Intrinsics Guide.)

I found that when the LLVM prefix was different, the intrinsics were actually part of a different target feature, but it wouldn’t surprise me if that is not consistent. The periods in intrinsic names are not consistent either. And sometimes the C intrinsics aren’t implemented in terms of LLVM intrinsics at all ([example](http://clang.llvm.org/doxygen/avxintrin_8h_source.html#l01605)), or in terms of `shufflevector` which I haven’t got a clue how to make that work from Rust.

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to resubmit with a rebase and comments addressed!
